src/main/scala/com/fulcrumgenomics/alignment/NeedlemanWunschAligner.scala

```diff
@@ -0,0 +1,187 @@
+    *
+    * Matrix is constructed such that:
+    *   - Matrix as nRow * nColumn (i.e. coordinates are (row, column))
+    *   - Query sequence is down the side (each row represents a query base)
+    *   - Target sequence is along the top (each column represents a target base)
+    *   - i is iterating over rows/query bases
+    *   - j is iterating over columns/target bases
+    *   - Up   == Consumes query base  but not target base, i.e. Insertion vs. the target
+    *   - Left == Consumes target base but not query base, i.e. Deletion vs. the target
+    *
+    * @param query the query sequence
+    * @param target the target sequence
+    * @return a tuple of the scoring matrix and the traceback matrix
+    */
+  protected def buildMatrices(query: Array[Byte], target: Array[Byte]): (Matrix[Int], Matrix[Direction]) = {
+    val scoring = Matrix[Int](query.length + 1, target.length + 1)
+    val trace   = Matrix[Direction](query.length + 1, target.length + 1)
+
+    // Top left corner
+    scoring(0,0) = 0
+    trace(0,0)   = Done
+
+    // Down the left hand side
+    forloop(from=1, until=query.length+1) { i =>
+      trace(i, 0)   = Up
+      scoring(i, 0) = scoring(i-1, 0) + gapExtend + (if (i == 1) gapOpen else 0)
+    }
+
+    // Along the top
+    forloop(from=1, until=target.length+1) { j =>
+      trace(0, j)   = Left
+      scoring(0, j) = scoring(0, j-1) + gapExtend + (if (j == 1) gapOpen else 0)
+    }
+
+    // The interior of the matrix
+    forloop(from=1, until=query.length+1) { i =>
+      forloop(from=1, until=target.length+1) { j =>
+        val diagonalScore = scoring(i-1, j-1) + scoringFunction(query(i-1), target(j-1))
+        val upScore       = scoring(i-1, j  ) + gapExtend + (if (trace(i-1, j  ) == Up) 0 else gapOpen)
+        val leftScore     = scoring(i,   j-1) + gapExtend + (if (trace(i,   j-1) == Left) 0 else gapOpen)
+
+        val maxScore = max(max(diagonalScore, upScore), leftScore)
+        scoring(i,j) = maxScore
+        trace(i,j)   = if (maxScore == diagonalScore) Diagonal else if (maxScore == leftScore) Left else Up
+      }
+    }
+
+    (scoring, trace)
+  }
+
+  /**
+    * Given the scoring and trace back matrices, construct the [[Alignment]] object for the (or one of the)
+    * optimal alignments.
+    *
+    * @param query the query sequence
+    * @param target the target sequence
+    * @param scoring the scoring matrix
+    * @param trace the trace back matrix
+    * @return an [[Alignment]] object representing the alignment
+    */
+  protected def generateAlignment(query: Array[Byte], target: Array[Byte], scoring: Matrix[Int], trace: Matrix[Direction]): Alignment = {
+    var i = query.length
+    var j = target.length
+    var currOperator: CigarOperator = null
+    var currLength: Int = 0
+    val stack = new mutable.Stack[CigarElem]()
+
+    while (i != 0 || j != 0) {
+      val op = trace(i,j) match {
+        case Done     =>
+          i = 0
+          j = 0
+          null
```

Comment:
Remove this case, as I don't think we will ever hit it, since `Done` is only stored in `trace(0, 0)`.  If you intend this to be used for other types of alignment (maybe that's why it's `protected`) ignore this comment, since then yeah, this is needed.